### PR TITLE
[#2] Setup CI

### LINKED ISSUE #2: Setup CI
## Why

To align with Nimble's workflow and ensure the effectiveness of the product we should have automated CI

## Acceptance Criteria

- [ ] Setup CI with Github Action

### PR DESCRIPTION
**Pull request by carryall**
Resolve https://github.com/carryall/react-survey-challenge/issues/2

## What happened 👀

Add CI workflow to Github Action
- Unit test
  - Linting
  - Jest test
- UI test
  - Cypress test

## Insight 📝

N/A

## Proof Of Work 📹

The workflow successfully run
- [Unit test](https://github.com/carryall/react-survey-challenge/actions/runs/1331724324)
- [UI test](https://github.com/carryall/react-survey-challenge/actions/runs/1331724325)

Keep the Cypress video in the artifact
Screen Shot 2564-10-12 at 12 20 26

### CODE REVIEW

File: .github/workflows/test-ui.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        if: failure()
+        with:
+          name: cypress-screenshots
+          path: cypress/screenshots
+
+      - uses: actions/upload-artifact@v2
+        if: always()
+        with:
+          name: cypress-videos
+          path: cypress/videos
```
Comment (carryall) [inline .github/workflows/test-ui.yml]: Store screenshot in artifact only when the test failed and always store the video

File: .github/workflows/test-ui.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+on:
+  push:
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x]
```
Comment (hoangmirs) [inline .github/workflows/test-ui.yml:12]: You declare matrix here, do we need to use it? I don't see it in the below code. Or will it be used automatically?

Comment (hoangmirs) [reply .github/workflows/test-ui.yml:12]: One more thing, I think we can use [setup-node](https://github.com/actions/setup-node#matrix-testing) action to install the needed node version, and it supports caching as well

File: .github/workflows/test-ui.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        run: yarn
+
+      - name: Run Cypress test
+        uses: cypress-io/github-action@v2
+        with:
+          start: yarn start
+          wait-on: http://localhost:3000
+          browser: chrome
+
+      - uses: actions/upload-artifact@v2
```
Comment (hoangmirs) [inline .github/workflows/test-ui.yml]: Are you missing names for these steps?

File: .github/workflows/test-ui.yml (hunk 2 above)
Comment (carryall) [reply .github/workflows/test-ui.yml:12]: Ahh I forgot to add the setup node step 😅 , added it in 58d74f3

File: .github/workflows/test-ui.yml (hunk 3 above)
Comment (carryall) [reply .github/workflows/test-ui.yml]: Thanks, added in 58d74f3

File: .github/workflows/test-ui.yml (hunk 2 above)
Comment (hoangmirs) [reply .github/workflows/test-ui.yml:12]: @carryall should we enable cache as well?

File: .github/workflows/test-ui.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        uses: styfle/cancel-workflow-action@0.9.1
+        with:
+          access_token: ${{ github.token }}
+
+      - name: Checkout
+        uses: actions/checkout@v2.3.4
+
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
+        with:
```
Comment (hoangmirs) [inline .github/workflows/test-ui.yml:25]: @carryall should we enable cache as well?

File: .github/workflows/test-ui.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        run: yarn
+
+      - name: Run Cypress test
+        uses: cypress-io/github-action@v2
+        with:
+          start: yarn start
+          wait-on: http://localhost:3000
+          browser: chrome
+
+      - name: Upload screenshots
```
Comment (hoangmirs) [inline .github/workflows/test-ui.yml]: Should we change to "Upload test run screenshots" or "test execution screenshots"?

File: .github/workflows/test-unit.yml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        uses: styfle/cancel-workflow-action@0.9.1
+        with:
+          access_token: ${{ github.token }}
+
+      - name: Checkout
+        uses: actions/checkout@v2.3.4
+
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
+        with:
```
Comment (hoangmirs) [inline .github/workflows/test-unit.yml:25]: Should we add cache as well?

File: .github/workflows/test-unit.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+          node-version: ${{ matrix.node-version }}
+
+      - name: Install dependency
```
Comment (hoangmirs) [inline .github/workflows/test-unit.yml]:
```suggestion
      - name: Install dependencies
```

File: .github/workflows/test-unit.yml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        run: yarn
+
+      - name: Run Codebase Lint
```
Comment (hoangmirs) [inline .github/workflows/test-unit.yml]: Should it be? 🤔 
```suggestion
      - name: Run codebase linter
```

**Review by hoangmirs — COMMENTED** [review]
I saw this section at the end of Pull request files changed. Do you also see it?
__2__Setup_CI_by_carryall_·_Pull_Request__27_·_carryall_react-survey-challenge

If yes, should we create another PR to fix it?

**Comment by carryall** [comment]
> I saw this section at the end of Pull request files changed. Do you also see it? 
> 
> If yes, should we create another PR to fix it?

Just noticed it Thanks @hoangmirs it's the warning from the linter, the code is from the template 🙈  
fixed it in 05c7872

File: .github/workflows/test-ui.yml (hunk 4 above)
Comment (carryall) [reply .github/workflows/test-ui.yml:25]: Good point I added the cache in bbe7aee

File: .github/workflows/test-ui.yml (hunk 5 above)
Comment (carryall) [reply .github/workflows/test-ui.yml]: Updated the name in b43a1e7 along with other comments about naming 🙌🏼

File: .github/workflows/test-ui.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+          access_token: ${{ github.token }}
+
+      - name: Checkout
+        uses: actions/checkout@v2.3.4
+
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: 'npm'
```
Comment (hoangmirs) [inline .github/workflows/test-ui.yml]: I see a warning on this line. The latest version of this action is `v2.4.1`, should we upgrade it.
__3__Cache_npm_dependencies_·_carryall_react-survey-challenge_c9ca4e0
 

And you are using yarn to manage dependencies, should we use `cache: 'yarn'` as well?
And please check all other actions and upgrade them to the latest version as well 🙇

Comment (carryall) [reply .github/workflows/test-ui.yml]: Thanks, @hoangmirs I updated all versions and cache yarn in 4fdd9c6 
the warning is now gone, you can check on [this](https://github.com/carryall/react-survey-challenge/actions/runs/1357459575) and [this](https://github.com/carryall/react-survey-challenge/actions/runs/1357459574) run 🙇🏼

**Review by hoangmirs — APPROVED** [review]

**Review by rafayet-monon — APPROVED** [review]

**Pull request merged**